components/renku_data_services/solr/solr_migrate.py

```diff
@@ -0,0 +1,209 @@
+    SolrClientConfig,
+)
+from renku_data_services.solr.solr_schema import (
+    AddCommand,
+    CopyFieldRule,
+    CoreSchema,
+    DeleteDynamicFieldCommand,
+    DeleteFieldCommand,
+    DeleteFieldTypeCommand,
+    DynamicFieldRule,
+    Field,
+    FieldType,
+    ReplaceCommand,
+    SchemaCommand,
+    SchemaCommandList,
+)
+
+
+def _is_applied(schema: CoreSchema, cmd: SchemaCommand) -> bool:
+    """Check whether a schema command is already applied to the given schema."""
+    match cmd:
+        case AddCommand(FieldType() as ft):
+            return any(x.name == ft.name for x in schema.fieldTypes)
+
+        case AddCommand(Field() as f):
+            return any(x.name == f.name for x in schema.fields)
+
+        case AddCommand(DynamicFieldRule() as f):
+            return any(x.name == f.name for x in schema.dynamicFields)
+
+        case AddCommand(CopyFieldRule() as f):
+            return any(x.source == f.source and x.dest == f.dest for x in schema.copyFields)
+
+        case DeleteFieldCommand(f):
+            return all(x.name != f for x in schema.fields)
+
+        case DeleteFieldTypeCommand(f):
+            return all(x.name != f for x in schema.fieldTypes)
+
+        case DeleteDynamicFieldCommand(f):
+            return all(x.name != f for x in schema.dynamicFields)
+
+        case ReplaceCommand(FieldType() as ft):
+            return any(x == ft for x in schema.fieldTypes)
+
+        case ReplaceCommand(Field() as f):
+            return any(x == f for x in schema.fields)
+
+        case _:
+            return False
+
+
+@dataclass
+class SchemaMigration:
+    """A migration consisting of the version and a set of schema commands."""
+
+    version: int
+    commands: list[SchemaCommand]
+    requires_reindex: bool
+
+    def is_empty(self) -> bool:
+        """Return whether the migration contains any commands."""
+        return self.commands == []
+
+    def align_with(self, schema: CoreSchema) -> Self:
+        """Aligns the list of schema commands to the given schema.
+
+        Return a copy of this value, removing all schema commands that have already
+        been applied to the given schema.
+        """
+        cmds = list(filter(lambda e: not (_is_applied(schema, e)), self.commands))
+        return type(self)(version=self.version, commands=cmds, requires_reindex=self.requires_reindex)
+
+
+@dataclass
+class MigrateResult:
+    """The overall result of running a set of migrations."""
+
+    startVersion: int | None
+    endVersion: int | None
+    migrationsRun: int
+    migrationsSkipped: int
+    requiresReindex: bool
+
+    @classmethod
+    def empty(cls) -> "MigrateResult":
+        """Create an empty MigrateResult."""
+        return MigrateResult(None, None, 0, 0, False)
+
+
+class VersionDoc(BaseModel):
+    """A document tracking the schema migration."""
+
+    id: str = "VERSION_ID_EB779C6B-1D96-47CB-B304-BECF15E4A607"
+    current_schema_version_l: int
+    migration_running_b: bool
```

Comment:
question: what does the `_b` and `_l` refer to in the names here?

Reply:
This is a solr specific thing. since this is the document that gets inserted before any of our schema migration runs, it uses solr dynamic fields. you append a `_<type-letter>` to a name. So a `_b` is a `bool` and a `_l` is a long/int.